Docs/subscriptions (#1437)
neo4j/graphql: docs/modules/ROOT/pages/subscriptions/scaling.adoc

[[horizontal-scaling]]
= Horizontal Scaling

WARNING: Subscriptions are only available as a beta, its API may change in the future. It is not recommended to use subscriptions in production environments.

> ## The Problem

Reviewer: Mixing and matching markdown and AsciiDoc in this file!

```suggestion
== The Problem
```

Horizontal scaling of any real time system can be tricky. Especially when dealing with long lived connections such as WebSockets.
Consider the following example, in which _Client A_ is subscribed to a certain event, that is triggered by _Client B_:

image::subscriptions/diagram1.png[title="Simple subscriptions setup"]

In the previous example, the server running the GraphQL service, will do the following:

1. Receive the _mutation_ by _Client B_.
2. Run the Cypher Query on Neo4j.
3. Trigger the subscription event to _Client A_.

This setup works fine for a single instance of `@neo4j/graphql`. However, when trying to scale horizontally, by adding more GraphQL servers,
we may encounter the following situation:

image::subscriptions/diagram2.png[title="Subscriptions with 2 servers"]

In this case, _Client B_ is subscribed to one server, however, when _Client A_ triggers the mutation, it queries a different server.

> The change happens successfully in the database, and any client connected to the same server will receive the subscription event, however, _Client A_

Reviewer 2: ```suggestion
The change happens successfully in the database, and any client connected to the same server will receive the subscription event. However, _Client A_
```

> will **not** receive any update, as the server its connected to will not get notified of any mutation.

Reviewer 2: ```suggestion
will **not** receive any update, as the server it's connected to will not get notified of any mutation.
```

> This is the behaviour of the xref::subscriptions/plugins/single-instance.adoc[single instance plugin] provided by the library, and it is not intended for production usage.

Reviewer: ```suggestion
This is the behaviour of the xref::subscriptions/plugins/single-instance.adoc[single instance plugin] provided by the library, so it is not suitable for use in a horizontally scaled environment.
```
